File: lista-encadeada/java/todolist/TaskList.java
package lists.linkedlist.todolist;

import java.util.ArrayList;
import java.util.List;

import lists.linkedlist.LinkedList;
import lists.linkedlist.Node;

public class TaskList {
    
    private LinkedList<Task> tasks;

    public TaskList() {
        this.tasks = new LinkedList<>();
    }

    public List<Task> getTasks() {
        return tasks.toArray();
    }

    public void addTask(Task task) {
        tasks.addAtEnd(task);
    }

    public void addTask(Task task, Integer index) {
        if (index == null) {
            tasks.addAtEnd(task);
            return;
        }
        if (tasks.get(index) == null) {
            return;
        }
        tasks.addAtPosition(index, task);
    }

    public List<Task> getTasksByTag(String tag) {
        Node<Task> current = tasks.getHead();
        List<Task> tasksByTag = new ArrayList<>();

        while (current != null) {
            if (current.getValue().getTag().equals(tag)) {
                tasksByTag.add(current.getValue());
            }
            current = current.getNext();
        }

        return tasksByTag;
    }

    public Task getTaskById(Integer id) {
        Node<Task> current = tasks.getHead();
        while (current != null) {
            if (current.getValue().getId() == id) {
                return current.getValue();
            }
            current = current.getNext();
        }
        return null;
    }

    public Task removeTaskById(Integer id) {
        Task taskItem = getTaskById(id);
        if (taskItem != null) {
            tasks.remove(taskItem);
        }
        return taskItem;
    }

    public Task setTaskToCompleted(Integer id) {
        Task taskItem = getTaskById(id);
        if (taskItem != null) {
            taskItem.setStatus(TaskStatus.COMPLETED);
        }
        return taskItem;
    }

    public Task setTaskDataById(Integer id, Task task) {
        Task taskItem = getTaskById(id);
        if (taskItem != null) {
            taskItem.setDescription(task.getDescription());
            taskItem.setStatus(task.getStatus());
            taskItem.setTag(task.getTag());
        }
        return taskItem;
    }

    public void move(Integer id, int targetIndex) {
        if (targetIndex < 0 || targetIndex >= tasks.getSize()) {
            return;
        }
        Task taskItem = getTaskById(id);
        int sourceIndex = tasks.indexOf(taskItem);
        if (taskItem != null && sourceIndex != targetIndex) {
            tasks.removeAtPosition(sourceIndex);
            tasks.addAtPosition(targetIndex, taskItem);
        }
    } 
}

File: lista-encadeada/csharp/todolist/TaskList.cs
using System;
using System.Collections.Generic;

namespace Lists.LinkedList.TodoList
{
    public class TaskList {
        private LinkedList<Task> tasks;

        public TaskList() {
            this.tasks = new LinkedList<Task>();
        }

        public List<Task> GetTasks() {
            return tasks.ToArray();
        }

        public void AddTask(Task task) {
            tasks.AddAtEnd(task);
        }

        public void AddTask(Task task, int? index) {
            if (index == null) {
                tasks.AddAtEnd(task);
                return;
            }
            if (tasks.Get(index.Value) == null) {
                return;
            }
            tasks.AddAtPosition(index.Value, task);
        }

        public List<Task> GetTasksByTag(string tag) {
            Node<Task> current = tasks.GetHead();
            List<Task> tasksByTag = new List<Task>();

            while (current != null) {
                if (current.Value.Tag.Equals(tag)) {
                    tasksByTag.Add(current.Value);
                }
                current = current.Next;
            }
            return tasksByTag;
        }

        public Task GetTaskById(int id) {
            Node<Task> current = tasks.GetHead();
            while (current != null) {
                if (current.Value.Id == id) {
                    return current.Value;
                }
                current = current.Next;
            }
            return null;
        }

        public Task RemoveTaskById(int id) {
            Task taskItem = GetTaskById(id);
            if (taskItem != null) {
                tasks.Remove(taskItem);
            }
            return taskItem;
        }

        public Task SetTaskToCompleted(int id) {
            Task taskItem = GetTaskById(id);
            if (taskItem != null) {
                taskItem.Status = TaskStatus.Completed;
            }
            return taskItem;
        }

        public Task SetTaskDataById(int id, Task task) {
            Task taskItem = GetTaskById(id);
            if (taskItem != null) {
                taskItem.Description = task.Description;
                taskItem.Status = task.Status;
                taskItem.Tag = task.Tag;
            }
            return taskItem;
        }

        public void Move(int id, int targetIndex) {
            if (targetIndex < 0 || targetIndex >= tasks.GetSize()) {
                return;
            }
            Task taskItem = GetTaskById(id);
            int sourceIndex = tasks.IndexOf(taskItem);
            if (taskItem != null && sourceIndex != targetIndex) {
                tasks.RemoveAtPosition(sourceIndex);
                tasks.AddAtPosition(targetIndex, taskItem);
            }
        }
    }
}
